File: fsharp/clock/Clock.fs
module Clock

type Clock =
  { Hours: int
    Minutes: int }

let processMinutes (minutes: int) : (int * int) =
    let minute = minutes % 60
    let hourOffset = minutes / 60

    if minute >= 0
        then (hourOffset, minute)
        else (hourOffset - 1, 60 + minute)

let processHours (hours: int) : int =
    let hour = hours % 24
    if hour >= 0
        then hour
        else 24 + hour

let create (hours:int) (minutes: int) =
    let (hourOffset, minute) = processMinutes minutes
    let hour = processHours (hours + hourOffset)

    { Hours = hour; Minutes = minute }

let add minutes clock =
   create clock.Hours (clock.Minutes + minutes)

let subtract minutes clock =
   create clock.Hours (clock.Minutes - minutes)

let display clock =
    sprintf "%02d:%02d" clock.Hours clock.Minutes

File: typescript/clock/clock.ts
const MINUTES_IN_DAY = 60 * 24
const MINUTES_IN_HOUR = 60

export class Clock {
  private readonly _minutes: number
  
  constructor(hour: number, minute: number = 0) {
    const minutes = (hour * MINUTES_IN_HOUR + minute) % MINUTES_IN_DAY
    this._minutes = minutes < 0 ? MINUTES_IN_DAY + minutes : minutes
  }

  private hours = () =>
    `${String(Math.trunc(this._minutes / MINUTES_IN_HOUR)).padStart(2, '0')}`

  private minutes = () => 
    `${String(this._minutes % MINUTES_IN_HOUR).padStart(2, '0')}`
  
  public toString = () =>
    `${this.hours()}:${this.minutes()}`

  public plus = (minutes: number): Clock =>
    new Clock(0, this._minutes + minutes)

  public minus = (minutes: number): Clock =>
    new Clock(0, this._minutes - minutes)

  public equals = (other: Clock): boolean =>
    this.toString() === other.toString()
}
